drivers: counter: mcux_lptmr: enhance idle timer support for cortex_m_systick

Holt-Sun opened the pull request:
## Overview

This PR enhances the NXP MCUX LPTMR counter driver to better support its use as a companion idle timer for the Cortex-M SysTick driver during low-power mode operation.

## Motivation

When using `CONFIG_CORTEX_M_SYSTICK_LPM_TIMER_COUNTER`, an LPTMR instance serves as the timekeeping source while SysTick is not clocked during low-power states. The current driver has limitations that require workarounds in the idle timer integration code:

1. **No free-running mode control**: The driver hardcodes `enableFreeRunning = false`, requiring code modification to maintain continuous timekeeping across multiple wake events
2. **No support for immediate alarms**: Setting `ticks=0` returns `-EINVAL`, forcing special-case handling when calculated sleep duration is minimal
3. **Incomplete pending status**: `get_pending_int()` doesn't detect manually-triggered interrupts

## Changes

### Patch 1: Add device tree support for free-running mode

Adds `run-mode` property to configure LPTMR reset behavior per instance:
- `restart` (default): Counter resets after compare match (backward compatible)
- `free-run`: Counter continues after compare, resets only on overflow

**Benefits:**
- Devicetree-based configuration without code changes
- Supports different operational modes for different instances
- Essential for continuous timekeeping in idle timer scenarios

**Example:**
```dts
&lptmr0 {
    status = "okay";
    run-mode = "free-run";  /* Don't reset on compare */
};
```

### Patch 2: Fire alarm immediately when ticks=0

Allows `counter_set_alarm()` with `ticks=0` to trigger immediate callback via interrupt pending flag.

**Implementation:**
- Remove validation that rejects `ticks == 0`
- Add `irq_set_pending()` and `irq_is_pending()` helpers supporting both NVIC and GIC interrupt controllers
- Store IRQ number in device config structure
- When `ticks == 0`, set interrupt pending to trigger ISR immediately
- Refactor timer state handling to eliminate code duplication
- Enhance `mcux_lptmr_get_pending_int()` to check both hardware flags and interrupt controller pending state

**Benefits:**
- Eliminates special-case handling for zero/minimal sleep durations
- Provides consistent Counter API behavior for all tick values

## Backward Compatibility

Both patches maintain full backward compatibility:
- Default `run-mode` is `restart` (existing behavior)
- Normal alarm operations (ticks > 0) are unchanged
- Existing devicetree configurations work without modification

Holt-Sun committed:
drivers: counter: mcux_lptmr: add device tree support for freerunning mode

Add device tree property to configure LPTMR free-running mode instead
of hardcoding it to disabled.

The freerunning property controls counter reset behavior:
- If false (default): Counter resets after compare match
- If true: Counter resets only on overflow, continues after compare

This allows per-instance configuration of the LPTMR timer mode based
on application requirements without code changes.

Signed-off-by: Holt Sun <holt.sun@nxp.com>

ZhaoxiangJin reviewed (dismissed):
ah, we back to enable this feature again. when lptmr is used as the systick companion low-power mode timer, enabling the free running feature can indeed improve compensation accuracy. In my opinion, adding this feature is meaningful.

Holt-Sun commented:
> Hi @Holt-Sun Could you also look at the GPT (General Purpose Timer) counter driver (RT1xxx)? As I remember, it has the same functional gap.

Thank you for reminder @butok. I will do it and in another PR possible.
Please firstly review this PR, thanks.

butok commented:
> > Not sure what the default state should be: "compare" or "free-run". As I remember, the GPT counter is initialized to free-run.
> 
> For GPT, hardware reset to restart mode. In our driver, hardcode to free-run mode:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/8e5c223b5454b4cacacea8bd533f769092623bbf/drivers/counter/counter_mcux_gpt.c#L208
> 
> This PR makes it configurable keeping backward compatibility: #101621

Thank you. 
But I believe it's better to unify the default value ("compare"/"free-run") and its method of modification in DT for both LPTMR and GPT.

butok commented:
>I agree that it is better to unify the method of modification in DT for both LPTMR and GPT. I have updated both PRs apply it.

Thank you.

> For default value, it is better to keep LPTMR’s default as restart (current behavior) to avoid silently changing behavior affects all existing project in tree or out of tree. The default for GPT is free-run to keep backward compatibility.

Guess, it was done unintentionally during the basic enablement.
I believe the default value should be the same, for consistency. 
No matter which one you choose - it's up to you.

Holt-Sun committed:
drivers: counter: mcux_lptmr: fire alarm immediately when ticks=0

Allow alarm_cfg->ticks to be 0 for immediate callback execution.
When ticks equals 0, set the interrupt pending instead of returning
-EINVAL, causing the ISR to fire immediately.

Implementation:
- Add irq_set_pending()/irq_is_pending() helpers for NVIC and GIC
- Store IRQ number in device config for pending operations
- Refactor timer state handling to eliminate duplication
- Update get_pending_int() to check both HW flags and NVIC/GIC state

This supports idle timer use cases where sleep duration may be zero
and eliminates the need for special-case handling in application code.

Signed-off-by: Holt Sun <holt.sun@nxp.com>

Albort12138 reviewed (commented):
> > Regarding the current update, it may not align with the definition of Free-running mode. Taking the alarm function as an example: Once the mcux_lptmr_set_alarm() function is invoked, if the LPTMR module is enabled, the function will first disable the LPTMR; stopping the timer will clear the CNR register. Additionally, within the mcux_lptmr_isr(), the LPTMR module is disabled—this action clears the CNR register even when operating in Free-running mode. This indicates that while the LPTMR hardware operates in free-running mode (where the timer does not reset to zero when the compare flag is asserted), the software implementation disables the LPTMR module and manually resets it to zero.
> 
> Here is the execution sequence:
> ```
> I: About to enter deep sleep
> PM >
> pm_system_suspend: entry
> pm_system_suspend: kernel_ticks=11002, events_ticks=-1
> pm_system_suspend: ticks=11002
> pm_system_suspend: selected state=2
> pm_system_suspend: exit_latency_ticks=5
> pm_system_suspend: will enter sys_clock_set_timeout
> sys_clock_set_timeout: entering low power mode with timeout_us=1099700
> z_cms_lptim_hook_on_lpm_entry: max_lpm_time_us=1099700 ticks=4398
> mcux_lptmr_set_alarm: ticks=4398
> mcux_lptmr_set_alarm: check pending int before setting alarm: pending=0
> 
> [21:54:02.482]_isr_wrapper: Enter pm_system_resume in _kernel.idle.
> pm_system_resume: entry
> pm_system_resume: post_ops_required=1
> pm_system_resume: doing post ops.
> pm_system_resume: calling sys_clock_idle_exit.
> sys_clock_idle_exit: calling z_cms_lptim_hook_on_lpm_exit
> z_cms_lptim_hook_on_lpm_exit: entered
> z_cms_lptim_hook_on_lpm_exit: directly reading idle timer value due to pending interrupt
> z_cms_lptim_hook_on_lpm_exit: idle_timer_pre=0 idle_timer_post=4545 pending=1 wrap=1 diff=8944 us=2236000
> z_cms_lptim_hook_on_lpm_exit: exiting
> sys_clock_idle_exit: systick_us=89425275 idle_timer_us=2236000
> sys_clock_idle_exit: missed_cycles=0 dticks=1028
> _isr_wrapper: Back from pm_system_resume in _kernel.idle.
> _isr_wrapper: calling irq_number=29
> In mcux_lptmr_isr.
> pm_system_resume: entry
> pm_system_resume: post_ops_required=0
> ```
> When the lptmr interrupt happens, instead of directly going to mcux_lptmr_isr, the program will firstly go into [pm_system_resume](https://github.com/nxp-upstream/zephyr/blob/faaabbb2bbcee54788c8f315640d2352c2097e98/arch/arm/core/cortex_m/isr_wrapper.c#L62)/sys_clock_idle_exit/z_cms_lptim_hook_on_lpm_exit to use  [counter_get_value](https://github.com/nxp-upstream/zephyr/blob/faaabbb2bbcee54788c8f315640d2352c2097e98/drivers/timer/cortex_m_systick.c#L175) to get the count, with free-running mode configurable and enabled, at this time point, the CNR will keep running then the  [counter_get_value](https://github.com/nxp-upstream/zephyr/blob/faaabbb2bbcee54788c8f315640d2352c2097e98/drivers/timer/cortex_m_systick.c#L175) will get the correct time point. Without free-running mode enabled, when the lptmr interrupt happens, the CNR will be reset immediately.
> The mcux_lptmr_isr will be invoked only when the pm_system_resume in isr_wrapper.c returned, then as you said, stop the lptmr reseting CNR, but it doesn't matter then.
> 
> So the free-running mode configuration and enablement is important for keep the time point caught correctly.  

Thank you for your explanation.
If my understanding is correct.
In the scenario where the lptmr is utilized as a standard counter, the free-running mode is unnecessary. Specifically, the mcux_lptmr_isr interrupt service routine is triggered immediately when the CNR=CMR, and it will be manually reset within the mcux_lptmr_isr.
Given that the free-running mode is only functional when the lptmr is employed as a "Systick companion low-power mode timer", it may not be necessary to expose this property in the devicetree.

```c
#if (CONFIG_CORTEX_M_SYSTICK_LPM_TIMER)
lptmr_config.enableFreeRunning = true;
#else
lptmr_config.enableFreeRunning = false;
#endif

Holt-Sun commented:
> > > Regarding the current update, it may not align with the definition of Free-running mode. Taking the alarm function as an example: Once the mcux_lptmr_set_alarm() function is invoked, if the LPTMR module is enabled, the function will first disable the LPTMR; stopping the timer will clear the CNR register. Additionally, within the mcux_lptmr_isr(), the LPTMR module is disabled—this action clears the CNR register even when operating in Free-running mode. This indicates that while the LPTMR hardware operates in free-running mode (where the timer does not reset to zero when the compare flag is asserted), the software implementation disables the LPTMR module and manually resets it to zero.
> > 
> > 
> > Here is the execution sequence:
> > ```
> > I: About to enter deep sleep
> > PM >
> > pm_system_suspend: entry
> > pm_system_suspend: kernel_ticks=11002, events_ticks=-1
> > pm_system_suspend: ticks=11002
> > pm_system_suspend: selected state=2
> > pm_system_suspend: exit_latency_ticks=5
> > pm_system_suspend: will enter sys_clock_set_timeout
> > sys_clock_set_timeout: entering low power mode with timeout_us=1099700
> > z_cms_lptim_hook_on_lpm_entry: max_lpm_time_us=1099700 ticks=4398
> > mcux_lptmr_set_alarm: ticks=4398
> > mcux_lptmr_set_alarm: check pending int before setting alarm: pending=0
> > 
> > [21:54:02.482]_isr_wrapper: Enter pm_system_resume in _kernel.idle.
> > pm_system_resume: entry
> > pm_system_resume: post_ops_required=1
> > pm_system_resume: doing post ops.
> > pm_system_resume: calling sys_clock_idle_exit.
> > sys_clock_idle_exit: calling z_cms_lptim_hook_on_lpm_exit
> > z_cms_lptim_hook_on_lpm_exit: entered
> > z_cms_lptim_hook_on_lpm_exit: directly reading idle timer value due to pending interrupt
> > z_cms_lptim_hook_on_lpm_exit: idle_timer_pre=0 idle_timer_post=4545 pending=1 wrap=1 diff=8944 us=2236000
> > z_cms_lptim_hook_on_lpm_exit: exiting
> > sys_clock_idle_exit: systick_us=89425275 idle_timer_us=2236000
> > sys_clock_idle_exit: missed_cycles=0 dticks=1028
> > _isr_wrapper: Back from pm_system_resume in _kernel.idle.
> > _isr_wrapper: calling irq_number=29
> > In mcux_lptmr_isr.
> > pm_system_resume: entry
> > pm_system_resume: post_ops_required=0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When the lptmr interrupt happens, instead of directly going to mcux_lptmr_isr, the program will firstly go into [pm_system_resume](https://github.com/nxp-upstream/zephyr/blob/faaabbb2bbcee54788c8f315640d2352c2097e98/arch/arm/core/cortex_m/isr_wrapper.c#L62)/sys_clock_idle_exit/z_cms_lptim_hook_on_lpm_exit to use  [counter_get_value](https://github.com/nxp-upstream/zephyr/blob/faaabbb2bbcee54788c8f315640d2352c2097e98/drivers/timer/cortex_m_systick.c#L175) to get the count, with free-running mode configurable and enabled, at this time point, the CNR will keep running then the  [counter_get_value](https://github.com/nxp-upstream/zephyr/blob/faaabbb2bbcee54788c8f315640d2352c2097e98/drivers/timer/cortex_m_systick.c#L175) will get the correct time point. Without free-running mode enabled, when the lptmr interrupt happens, the CNR will be reset immediately.
> > The mcux_lptmr_isr will be invoked only when the pm_system_resume in isr_wrapper.c returned, then as you said, stop the lptmr reseting CNR, but it doesn't matter then.
> > So the free-running mode configuration and enablement is important for keep the time point caught correctly.
> 
> Thank you for your explanation. If my understanding is correct. In the scenario where the lptmr is utilized as a standard counter, the free-running mode is unnecessary. Specifically, the mcux_lptmr_isr interrupt service routine is triggered immediately when the CNR=CMR, and it will be manually reset within the mcux_lptmr_isr. Given that the free-running mode is only functional when the lptmr is employed as a "Systick companion low-power mode timer", it may not be necessary to expose this property in the devicetree.
> 
> ```c
> #if (CONFIG_CORTEX_M_SYSTICK_LPM_TIMER)
> lptmr_config.enableFreeRunning = true;
> #else
> lptmr_config.enableFreeRunning = false;
> #endif
> ```

I don't agree with the point "In the scenario where the lptmr is utilized as a standard counter, the free-running mode is unnecessary.". In cortex_m_systick.c, the lptmr is already utilized as a standard counter. We cannot make the conclusion that the free-running mode is only used for a standard companion counter for systick. Make it configurable with default "restart" value is much better and regular way than using `#if (CONFIG_CORTEX_M_SYSTICK_LPM_TIMER)` such guard macro which indicate the free-running mode configuration is only tight to this systick case.

Albort12138 commented:
> I don't agree with the point "In the scenario where the lptmr is utilized as a standard counter, the free-running mode is unnecessary.". In cortex_m_systick.c, the lptmr is already utilized as a standard counter. We cannot make the conclusion that the free-running mode is only used for a standard companion counter for systick. Make it configurable with default "restart" value is much better and regular way than using `#if (CONFIG_CORTEX_M_SYSTICK_LPM_TIMER)` such guard macro which indicate the free-running mode configuration is only tight to this systick case.

But when PM is not enabled (ISR is invoked as soon as an interrupt occurs), the free - running mode works just like the normal mode.

ZhaoxiangJin commented on drivers/counter/counter_mcux_lptmr.c:
```diff
@@ -98,18 +125,30 @@ static int mcux_lptmr_set_alarm(const struct device *dev, uint8_t chan_id,
 		/* RM recommendation: clear status flag after setting period
 		 * when timer is disabled.
 		 */
 		LPTMR_ClearStatusFlags(config->base, kLPTMR_TimerCompareFlag);
```
This code should placed after [LPTMR_SetTimerPeriod](https://github.com/zephyrproject-rtos/zephyr/pull/101573/changes/d67148eede755c82a434fbe78c052878a086081c#diff-70a99bb4a4583d7dbdf0c2a6c7397e0b98f23bb924fb680332f110e0d8d56f9dR151)? I am referring MCXN236 RM, it said: `When updating CMR, you must write to it and clear CSR[TCF] before the LPTMR counter increments past the new LPTMR compare value.`

FelixWang47831 commented on drivers/counter/counter_mcux_lptmr.c:
```diff
@@ -358,11 +413,13 @@ static DEVICE_API(counter, mcux_lptmr_driver_api) = {
 		.base = (LPTMR_Type *)DT_INST_REG_ADDR(n),			\
 		.clk_source = DT_INST_PROP(n, clk_source),			\
 		.bypass_prescaler_glitch = MCUX_LPTMR_BYPASS(n),	\
+		.freerunning = MCUX_LPTMR_IS_FREE_RUN(n),			\
```
Better to use snake case "free_running" to follow Zephyr coding style.

Holt-Sun commented:
> @Holt-Sun how were the different changes tested?

Hi @mmahadevan108 , I test 2 boards locally, unittests all build passed:
```
[09:56:57.609]*** Booting Zephyr OS build v4.3.0-3044-g8df62fe082b0 ***
Running TESTSUITE counter_basic
===================================================================

[09:56:57.923]START - test_all_channels
Testing lptmr@4a000

[09:56:58.057] PASS - test_all_channels in 0.133 seconds
===================================================================
START - test_cancelled_alarm_does_not_expire
Skipped for lptmr@4a000

[09:56:58.173] SKIP - test_cancelled_alarm_does_not_expire in 0.103 seconds
===================================================================
START - test_late_alarm
Skipped for lptmr@4a000

[09:56:58.289] SKIP - test_late_alarm in 0.103 seconds
===================================================================
START - test_late_alarm_error
Skipped for lptmr@4a000

[09:56:58.404] SKIP - test_late_alarm_error in 0.103 seconds
===================================================================
START - test_multiple_alarms
Skipped for lptmr@4a000

[09:56:58.518] SKIP - test_multiple_alarms in 0.103 seconds
===================================================================
START - test_set_top_value_with_alarm
Skipped for lptmr@4a000

[09:56:58.634] SKIP - test_set_top_value_with_alarm in 0.103 seconds
===================================================================
START - test_short_relative_alarm
Testing lptmr@4a000

[09:56:58.770] PASS - test_short_relative_alarm in 0.123 seconds
===================================================================
START - test_single_shot_alarm_notop
Testing lptmr@4a000

[09:56:58.956] PASS - test_single_shot_alarm_notop in 0.172 seconds
===================================================================
START - test_single_shot_alarm_top
Skipped for lptmr@4a000

[09:56:59.072] SKIP - test_single_shot_alarm_top in 0.103 seconds
===================================================================
START - test_valid_function_without_alarm
Skipped for lptmr@4a000

[09:56:59.189] SKIP - test_valid_function_without_alarm in 0.103 seconds
===================================================================
TESTSUITE counter_basic succeeded
Running TESTSUITE counter_no_callback
===================================================================

[09:56:59.513]START - test_set_top_value_without_alarm
Skipped for lptmr@4a000

[09:56:59.619] SKIP - test_set_top_value_without_alarm in 0.103 seconds
===================================================================
TESTSUITE counter_no_callback succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [counter_basic]: pass = 3, fail = 0, skip = 7, total = 10 duration = 1.149 seconds
 - PASS - [counter_basic.test_all_channels] duration = 0.133 seconds
 - SKIP - [counter_basic.test_cancelled_alarm_does_not_expire] duration = 0.103 seconds
 - SKIP - [counter_basic.test_late_alarm] duration = 0.103 seconds
 - SKIP - [counter_basic.test_late_alarm_error] duration = 0.103 seconds
 - SKIP - [counter_basic.test_multiple_alarms] duration = 0.103 seconds
 - SKIP - [counter_basic.test_set_top_value_with_alarm] duration = 0.103 seconds
 - PASS - [counter_basic.test_short_relative_alarm] duration = 0.123 seconds
 - PASS - [counter_basic.test_single_shot_alarm_notop] duration = 0.172 seconds
 - SKIP - [counter_basic.test_single_shot_alarm_top] duration = 0.103 seconds
 - SKIP - [counter_basic.test_valid_function_without_alarm] duration = 0.103 seconds

SUITE SKIP -   0.00% [counter_no_callback]: pass = 0, fail = 0, skip = 1, total = 1 duration = 0.103 seconds
 - SKIP - [counter_no_callback.test_set_top_value_without_alarm] duration = 0.103 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```
